serialize/decoder_test.go

```diff
@@ -0,0 +1,205 @@
+	"testing"
+
+	"github.com/m3db/m3db/x/xpool"
+	"github.com/m3db/m3x/checked"
+
+	"github.com/stretchr/testify/require"
+)
+
+const (
+	testCheckedBytesWrapperPoolSize = 2
+)
+
+var checkedBytesWrapperPool xpool.CheckedBytesWrapperPool
+
+func init() {
+	checkedBytesWrapperPool = xpool.NewCheckedBytesWrapperPool(testCheckedBytesWrapperPoolSize)
+	checkedBytesWrapperPool.Init()
+}
+
+func TestEmptyDecode(t *testing.T) {
+	var b []byte
+	b = append(b, headerMagicBytes...)
+	b = append(b, []byte{0x0, 0x0}...)
+
+	d := newDecoder(nil, checkedBytesWrapperPool)
+	d.Reset(wrapAsCheckedBytes(b))
+	require.False(t, d.Next())
+	require.NoError(t, d.Err())
+	d.Close()
+	d.Finalize()
+}
+
+func TestDecodeHeaderMissing(t *testing.T) {
+	var b []byte
+	b = append(b, []byte{0x0, 0x0}...)
+	b = append(b, []byte{0x0, 0x0}...)
+
+	d := newTestDecoder()
+	d.Reset(wrapAsCheckedBytes(b))
+	require.False(t, d.Next())
+	require.Error(t, d.Err())
+	d.Close()
+	d.Finalize()
+}
+
+func TestDecodeSimple(t *testing.T) {
+	b := testDecoderBytes()
+	d := newTestDecoder()
+	d.Reset(b)
+	require.NoError(t, d.Err())
+
+	require.True(t, d.Next())
+	tag := d.Current()
+	require.Equal(t, "abc", tag.Name.String())
+	require.Equal(t, "defg", tag.Value.String())
+
+	require.True(t, d.Next())
+	tag = d.Current()
+	require.Equal(t, "x", tag.Name.String())
+	require.Equal(t, "bar", tag.Value.String())
+
+	require.False(t, d.Next())
+	require.NoError(t, d.Err())
+
+	d.Close()
+	d.Finalize()
+}
+
+func TestDecodeMissingTags(t *testing.T) {
+	var b []byte
+	b = append(b, headerMagicBytes...)
+	b = append(b, encodeUInt16(uint16(2))...) /* num tags */
+
+	d := newTestDecoder()
+	d.Reset(wrapAsCheckedBytes(b))
+	require.NoError(t, d.Err())
+
+	require.False(t, d.Next())
+	require.Error(t, d.Err())
+}
+
+func TestDecodeOwnershipFinalize(t *testing.T) {
+	var b []byte
+	b = append(b, headerMagicBytes...)
+	b = append(b, encodeUInt16(uint16(2))...) /* num tags */
+
+	wrappedBytes := wrapAsCheckedBytes(b)
+	require.Equal(t, 0, wrappedBytes.NumRef())
+
+	d := newTestDecoder()
+	d.Reset(wrappedBytes)
+	require.NoError(t, d.Err())
+	require.NotEqual(t, 0, wrappedBytes.NumRef())
+
+	require.False(t, d.Next())
+	require.Error(t, d.Err())
+
+	d.Finalize()
+	require.Equal(t, 0, wrappedBytes.NumRef())
+	wrappedBytes.IncRef()
+	require.Nil(t, wrappedBytes.Get())
+}
+
+func TestDecodeMissingValue(t *testing.T) {
+	var b []byte
+	b = append(b, headerMagicBytes...)
+	b = append(b, encodeUInt16(uint16(2))...) /* num tags */
+	b = append(b, encodeUInt16(3)...)         /* len abc */
+	b = append(b, []byte("abc")...)
+
+	b = append(b, encodeUInt16(4)...) /* len defg */
+	b = append(b, []byte("defg")...)
+
+	b = append(b, encodeUInt16(1)...) /* len x */
+	b = append(b, []byte("x")...)
+
+	d := newTestDecoder()
+	d.Reset(wrapAsCheckedBytes(b))
+	require.NoError(t, d.Err())
+
+	require.True(t, d.Next())
+	require.False(t, d.Next())
+	require.Error(t, d.Err())
+}
+func TestDecodeCloneLifecycle(t *testing.T) {
```

Comment:
Can we perhaps add a check in this test that the checked bytes gets returned to pool somehow?

Reply:
Sure, added a couple of mock based tests and a SUT prop test for `Decoder` lifecycle guarantees.